Fix edge drawing performance regression

rossbar opened the pull request:
#4360 switched all edge drawing from LineCollection (in some cases) to always using FancyArrowPatches. This enabled nice new features such as self-loop drawing #4370 and a consistent, iterable return type, which makes it easier for users to modify individual edges (as FancyArrowPatches) after they have been created by `draw_networkx_edges`.

One downside of removing LineCollection is a large performance regression for drawing edges in undirected graphs, see #4806.

This PR tries to get the best of both worlds by re-introducing LineCollection for faster edge drawing, while keeping all the nice features of #4370 including drawing self-loops and the ability to draw curved edges.

The basic approach is:
 1. Add back the LineCollection drawing functionality that was removed in #4360.
 2. Encapsulate the two different drawing methods in internal functions
 3. Investigate how to trigger the different drawing methods based on existing kwargs, e.g. `arrows`, `arrowstyle`, `connectionstyle` etc.

rossbar commented:
Okay, I think this is close.

The toggling between FancyArrowPatch and LineCollection is the same as it was prior to #4360 - if a graph is directed and `arrows=True` (the default), then fancy arrow patches are used.

The current PR also ensures self-loops are always drawn by adding an extra draw step to specially handle the selfloops in the case where G is undirected.

rossbar commented:
I went ahead and requested a couple reviews to see what others think of this approach. It'd be good to know if there were any changes/improvements or other ideas that we should try before the release candidate comes out to prevent a plotting performance regression from 2.5 -> 2.6.

dschult commented:
Nice! -- and it gives good flexibility...

I think we should try to make a way for users to choose linecollection vs fancyarrows beyond the default system (which seems like fine defaults).  We've got the machinery, so it'd be good to let people use it even in nonstandard use-cases.

Could we enable arrows to be True/False/None... where None gives the default, False forces LineCollection and True forces FancyArrowPatch?  Something like:
```
    arrows : bool or None, optional (default=None)
        If None, directed graphs draw arrowheads returning FancyArrowPatches,
        while undirected graphs draw lines returning LineCollections for speed.
        If True, draw arrowheads using FancyArrowPatches (bendable and stylish).
        If False, draw lines using LineCollection (straight and fast).

Note: Arrowheads will be the same color as edges.
```
As for code, I think this means:
```python
    # Use LineCollection to draw edges by default, for performance reasons
    use_linecollection = G.is_directed()
    if arrows is True:
        use_linecollection = False
    if arrows is False:
        use_linecollection = True
```

dschult commented:
Nice! -- and it gives good flexibility...

I think we should try to make a way for users to choose linecollection vs fancyarrows beyond the default system (which seems like fine defaults).  We've got the machinery, so it'd be good to let people use it even in nonstandard use-cases.

Could we enable arrows to be True/False/None... where None gives the default, False forces LineCollection and True forces FancyArrowPatch?  Something like:
```
    arrows : bool or None, optional (default=None)
        If None, directed graphs draw arrowheads returning FancyArrowPatches,
        while undirected graphs draw lines returning LineCollections for speed.
        If True, draw arrowheads using FancyArrowPatches (bendable and stylish).
        If False, draw lines using LineCollection (straight and fast).

Note: Arrowheads will be the same color as edges.
```
As for code, I think this means:
```python
    # Use LineCollection to draw edges by default, for performance reasons
    use_linecollection = G.is_directed()
    if arrows is True:
        use_linecollection = False
    if arrows is False:
        use_linecollection = True
```

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -683,122 +681,155 @@ def draw_networkx_edges(
+        if edgelist:
+            edge_pos = np.asarray([(pos[e[0]], pos[e[1]]) for e in edgelist])
+            arrowstyle = "-"
+            _draw_networkx_edges_fancy_arrow_patch()
```
Should the FancyArrowPatch arrow_collection created for self-loops with undirected graphs (which is just a list if I read correctly)  get returned somehow?
I don't think it can be easily combined with LineCollection...  but I don't know much about combining the parts of a matplotlib drawing. It **does** get added to `ax` as a patch. So maybe that is enough.

I imagine not very many people will need to access those self-loops directly, but we provide that access for the other edges. Can we find a way to do it for these?

rossbar replied:
I had thought about this as well - one of the initial issues was about how different objects get returned based on the the inputs... We could in principle homogenize the output so that it always returns a tuple, e.g. `return (LineCollection, FancyArrowPatches)` where one or the other inputs may be empty depending on whether they were used in the drawing or not. If this were a brand new function, we could consider that; but since this change would break backward compatibility I don't think it's a good idea.

IMO the best way to access the edges to modify them after they've been created by `draw_networkx_edges` is through the `Axes` attributes like you've pointed out, e.g.

```python
ax = plt.gca()
ax.patches  # FancyArrowPatches for directed edges would can be accessed in this list
ax.collections  # LineCollections for undirected edges can be accessed here.
```

My preference would be to leave the current return types as-is and maybe just add a note to the documentation about how to access the various objects representing various types of edges.

dschult reviewed (commented):
The code looks good!  
I put some comments -- more like discussion points.
And also see my previous remark about preferring an option for people to turn on and off FancyArrowPatch for the non-default usage.

I think this is close to merging. :}

rossbar commented:
I like the idea of the 3-way switch on `arrows` proposed in [the above comment](https://github.com/networkx/networkx/pull/4825#issuecomment-847949481) -  I've implemented your suggestion (b33005f) and added a test for the logic (b487c76).

Re: the comment about whether to return the FAP's associated with self-loops, I went ahead and left the behavior as-is but added info about this situation to the `Notes` and `Examples` section of the docstring. I've also updated the `Returns` section and the `arrows` parameter description.

dschult commented on networkx/drawing/nx_pylab.py:
```diff
@@ -597,8 +599,11 @@ def draw_networkx_edges(
     -------
-    list of matplotlib.patches.FancyArrowPatch
-        `FancyArrowPatch` instances of the directed edges
+    list of matplotlib.patches.FancyArrowPatch or matplotlib.colections.LineCollection
+        If ``arrows=True``, a list of FancyArrowPatches is returned.
+        If 1`arrows=False``, a LineCollection is returned.
+        If ``arrows=None`` (the default), then a LineCollection is returned if
+        `G` is undirected, otherwise returns a list of FancyArrowPatches.
```
```suggestion
     matplotlib.colections.LineCollection or a list of matplotlib.patches.FancyArrowPatch
        If ``arrows=True``, a list of FancyArrowPatches is returned.
        If ``arrows=False``, a LineCollection is returned.
        If ``arrows=None`` (the default), then a LineCollection is returned if
        `G` is undirected, otherwise returns a list of FancyArrowPatches.
```
1 typo and maybe switch order of first line so it is clearly not a list of `LineCollection`s.

dschult reviewed (approved):
This looks good to me! 
The comment below is a typo and minor switch of words.
I approve and ready to merge.
